tooling/metamodel-generator/src/main/java/org/hibernate/processor/annotation/AnnotationMeta.java

```diff
@@ -136,24 +139,23 @@ && getAnnotationValue( mirror, "resultClass" ) == null ) {
 		}
 	}
 
-	private static @Nullable String resultType(SqmSelectStatement<?> selectStatement) {
-		final JpaSelection<?> selection = selectStatement.getSelection();
-		if (selection == null) {
-			return null;
-		}
-		else if (selection instanceof SqmSelectClause from) {
-			return from.getSelectionItems().size() > 1
-					? "Object[]"
-					: from.getSelectionItems().get(0).getJavaTypeName();
-		}
-		else if (selection instanceof JpaRoot<?> root) {
-			return root.getModel().getTypeName();
-		}
-		else if (selection instanceof JpaEntityJoin<?, ?> join) {
-			return join.getModel().getTypeName();
+	private @Nullable String resultType(SqmSelectStatement<?> selectStatement) {
+		final JavaType<?> javaType = selectStatement.getSelection().getJavaTypeDescriptor();
+		if ( javaType != null ) {
+			return javaType.getTypeName();
 		}
 		else {
-			return selection.getJavaTypeName();
+			final List<SqmSelectableNode<?>> items =
+					selectStatement.getQuerySpec().getSelectClause().getSelectionItems();
+			final SqmExpressible<?> expressible;
+			if ( items.size() == 1 && ( expressible = items.get( 0 ).getExpressible() ) != null ) {
+				final String typeName = expressible.getTypeName();
+				final TypeElement entityType = entityType( typeName );
+				return entityType == null ? typeName : entityType.getQualifiedName().toString();
+			}
+			else {
+				return "Object[]";
+			}
```

Comment:
This looks like it's copy/pasted from `NamedQueryMethod.returnType()`. Can we find some place to move the method and eliminate the duplication, please?

Reply:
That was the question I did not found the answer (maybe also reason why `entityType` method has not been moved). Because some SQM classes are used `Context` did not looked good can , and it can not be moved to `Sqm*` because of `Context` which is only used in processor.
As temporary solution I will create interface with default method that returns result type and abstract `getContext()`. This will work until better solution is found.

Reply:
`TypeUtil`?

Reply:
Maybe, if it is OK that some referencing SQMimports will be added
```
import org.hibernate.query.sqm.SqmExpressible;
import org.hibernate.query.sqm.tree.select.SqmSelectStatement;
import org.hibernate.query.sqm.tree.select.SqmSelectableNode;
```

Reply:
`SqmTypeUtils` sounds OK?

Reply:
Sure 👍